docs/source/getting-started/triggers.rst

```diff
@@ -0,0 +1,265 @@
+any.
+
+In order to start with something reasonably simple, we're going to set the rule as
+
+  *if* the last message we can see was *not* sent by ``bob``, *then* we'll send
+  ``"Please, tell me more about that."`` to whoever sent the last message we
+  can see.
+
+This raises the question of how we can determine which message is the last one,
+given the current structure of a message. In order to solve that, we need to
+add a ``Time`` field to ``Message``, which can be done by editing the
+``Message`` template in ``daml/User.daml`` to look like::
+
+    template Message with
+      sender: Party
+      receiver: Party
+      content: Text
+      receivedAt: Time
+    where
+      signatory sender, receiver
+
+This should result in Daml Studio reporting an error in the ``SendMessage``
+choice, as it now needs to set the ``receivedAt`` field. Here is the updated
+code for ``SendMessage``::
+
+        nonconsuming choice SendMessage: ContractId Message with
+            sender: Party
+            content: Text
+          controller sender
+          do
+            assertMsg "Designated user must follow you back to send a message" (elem sender following)
+            now <- getTime
+            create Message with sender, receiver = username, content, receivedAt = now
+
+The ``getTime`` action (`doc </daml/stdlib/Prelude.html#function-da-internal-lf-gettime-99334>`_)
+returns the time at which the command was received by the sandbox, which should
+be good enough for our purposes, though bear in mind that it may not be
+suitable for some use-cases.
+
+Now that we have a field to sort on, and thus a way to identify the *latest*
+message, we can turn our attention back to our trigger code.
+
+AutoReply
+*********
+
+Open up the trigger code again (``daml/ChatBot.daml``), and change it to::
+
+    module ChatBot where
+
+    import qualified Daml.Trigger as T
+    import qualified User
+    import qualified DA.List.Total as List
+
+    autoReply : T.Trigger ()
+    autoReply = T.Trigger
+      { initialize = pure ()
+      , updateState = \_ -> pure ()
+      , rule = \p -> do
+          message_contracts <- T.query @User.Message
+          let messages = map snd message_contracts
+          debug $ "Messages so far: " <> show (length messages)
+          let lastMessage = List.last $ List.sortOn (.receivedAt) messages
+          debug $ "Last message: " <> show lastMessage
+          case lastMessage of
+            Some m ->
+              if m.receiver == p
+              then do
+                users <- T.query @User.User
+                debug users
+                let isSender = (\user -> user.username == m.sender)
+                let replyTo = List.head $ filter (\(_, user) -> isSender user) users
+                case replyTo of
+                  None -> pure ()
+                  Some (sender, _) -> do
+                    T.dedupExercise sender (User.SendMessage p "Please, tell me more about that.")
```

Comment:
This dedup is fairly unreliable/slightly broken. Consider the following steps:

1. We get message M1 and respond here by sending a command asynchronously. Now we dedup on `(sender, SendMessage p "Please, tell me more about that.")`.
2. Now we get message M2 with a later time. We haven’t received a completion for M1 yet so dedup kicks in and the message gets droped.
3. Now we get a completion for M1 and dedup stops.

Arguably humans are slow at typing so the practical impacts of this are probably small but I don’t really like the idea of using a broken example in the GSG.